File: backend/services/chat_history_service.py
from datetime import datetime
from lib.mongodb import get_database
from models.chat import Chat, ChatMessage
from bson import ObjectId
from typing import List, Optional

class ChatHistoryService:
    async def get_chat_history_for_user(self, user_id: str) -> List[Chat]:
        db = get_database()
        chats_cursor = db.get_collection("chats").find({"userId": user_id}).sort("createdAt", -1)
        chats = []
        async for chat in chats_cursor:
            # Handle backward compatibility
            if "_id" in chat:
                chat["_id"] = str(chat["_id"])
            chats.append(Chat(**chat))
        return chats

    async def get_chat_by_id(self, chat_id: str) -> Optional[Chat]:
        db = get_database()
        chat = await db.get_collection("chats").find_one({"_id": ObjectId(chat_id)})
        if chat:
            # Handle backward compatibility
            if "_id" in chat:
                chat["_id"] = str(chat["_id"])
            return Chat(**chat)
        return None

    async def create_chat(self, user_id: str, name: str, agent_id: Optional[str] = None, model_id: Optional[str] = None) -> Chat:
        """Create a new chat with agent_id (preferred) or model_id (legacy)"""
        db = get_database()
        
        chat_data = {
            "userId": user_id,
            "name": name,
            "messages": [],
            "isPinned": False,
            "createdAt": datetime.utcnow(),
            "updatedAt": datetime.utcnow()
        }
        
        # Prefer agent_id over model_id
        if agent_id:
            chat_data["agentId"] = agent_id
        elif model_id:
            chat_data["modelId"] = model_id
        
        result = await db.get_collection("chats").insert_one(chat_data)
        
        # Create Chat object
        chat_data["_id"] = str(result.inserted_id)
        return Chat(**chat_data)

    async def create_chat_legacy(self, user_id: str, name: str, model_id: str) -> Chat:
        """Legacy method for backward compatibility"""
        return await self.create_chat(user_id, name, model_id=model_id)

    async def add_message_to_chat(self, chat_id: str, message: ChatMessage) -> Optional[Chat]:
        db = get_database()
        result = await db.get_collection("chats").find_one_and_update(
            {"_id": ObjectId(chat_id)},
            {
                "$push": {"messages": message.model_dump()}, 
                "$set": {"updatedAt": datetime.utcnow()}
            },
            return_document=True
        )
        if result and "_id" in result:
            result["_id"] = str(result["_id"])
            return Chat(**result)
        return None
        
    async def delete_chat(self, chat_id: str) -> bool:
        db = get_database()
        result = await db.get_collection("chats").delete_one({"_id": ObjectId(chat_id)})
        return result.deleted_count > 0

    async def update_chat_agent(self, chat_id: str, agent_id: str) -> Optional[Chat]:
        """Update chat to use new agent_id and remove legacy modelId"""
        db = get_database()
        result = await db.get_collection("chats").find_one_and_update(
            {"_id": ObjectId(chat_id)},
            {
                "$set": {
                    "agentId": agent_id,
                    "updatedAt": datetime.utcnow()
                },
                "$unset": {"modelId": "", "collectionNames": ""}
            },
            return_document=True
        )
        if result and "_id" in result:
            result["_id"] = str(result["_id"])
            return Chat(**result)
        return None

chat_history_service = ChatHistoryService() 

File: backend/src/chat/chat-history.service.ts
import { Injectable } from '@nestjs/common';
import { InjectModel } from '@nestjs/mongoose';
import { Model, Types } from 'mongoose';
import { Chat, ChatDocument } from '../models/chat.model';
import { ChatMessage } from '../models/chat.model';
// เพิ่ม import สำหรับ collection model
import { Collection, CollectionDocument } from '../models/collection.model';

@Injectable()
export class ChatHistoryService {
  constructor(
    @InjectModel(Chat.name) private chatModel: Model<ChatDocument>,
    @InjectModel('Collection') private collectionModel: Model<CollectionDocument>, // เพิ่ม collectionModel
  ) {}

  async getChatHistoryForUser(userId: string): Promise<Chat[]> {
    try {
      const chats = await this.chatModel
        .find({ userId })
        .sort({ updatedAt: -1 })
        .exec();
      return chats;
    } catch (error) {
      throw new Error(`Failed to get chat history for user: ${error.message}`);
    }
  }

  async getChatById(chatId: string): Promise<Chat | null> {
    try {
      const chat = await this.chatModel.findById(chatId).exec();
      return chat;
    } catch (error) {
      throw new Error(`Failed to get chat by ID: ${error.message}`);
    }
  }

  async createChat(
    userId: string,
    name: string,
    agentId?: string,
    modelId?: string,
    collectionNames?: string[],
    systemPrompt?: string,
    temperature?: number,
    maxTokens?: number,
  ): Promise<Chat> {
    try {
      const chatData: any = {
        // robust: แปลง userId เป็น ObjectId เสมอ
        userId: Types.ObjectId.isValid(userId) ? new Types.ObjectId(userId) : userId,
        name,
        title: name || 'New Chat',
        messages: [],
        isPinned: false,
        createdAt: new Date(),
        updatedAt: new Date(),
      };

      if (agentId) {
        chatData.agentId = Types.ObjectId.isValid(agentId) ? new Types.ObjectId(agentId) : agentId;
      }
      if (modelId) chatData.modelId = modelId;
      if (systemPrompt) chatData.systemPrompt = systemPrompt;
      if (typeof temperature === 'number') chatData.temperature = temperature;
      if (typeof maxTokens === 'number') chatData.maxTokens = maxTokens;

      // robust: validate collectionNames
      if (collectionNames && collectionNames.length > 0) {
        const collections = await this.collectionModel.find({ name: { $in: collectionNames } }).exec();
        if (collections.length !== collectionNames.length) {
          throw new Error('One or more collections not found');
        }
        chatData.collectionNames = collectionNames;
      }

      const chat = new this.chatModel(chatData);
      const savedChat = await chat.save();
      return savedChat.toJSON();
    } catch (error) {
      throw new Error(`Failed to create chat: ${error.message}`);
    }
  }

  async addMessageToChat(chatId: string, message: ChatMessage): Promise<Chat | null> {
    try {
      const result = await this.chatModel.findByIdAndUpdate(
        chatId,
        {
          $push: { messages: message },
          $set: { updatedAt: new Date() },
        },
        { new: true }
      ).exec();
      return result;
    } catch (error) {
      throw new Error(`Failed to add message to chat: ${error.message}`);
    }
  }

  async deleteChat(chatId: string): Promise<boolean> {
    try {
      const result = await this.chatModel.findByIdAndDelete(chatId).exec();
      return !!result;
    } catch (error) {
      throw new Error(`Failed to delete chat: ${error.message}`);
    }
  }

  async updateChatName(chatId: string, name: string): Promise<Chat | null> {
    try {
      const result = await this.chatModel.findByIdAndUpdate(
        chatId,
        {
          $set: { name, updatedAt: new Date() },
        },
        { new: true }
      ).exec();
      return result;
    } catch (error) {
      throw new Error(`Failed to update chat name: ${error.message}`);
    }
  }

  async updateChatPinStatus(chatId: string, isPinned: boolean): Promise<Chat | null> {
    try {
      const result = await this.chatModel.findByIdAndUpdate(
        chatId,
        {
          $set: { isPinned, updatedAt: new Date() },
        },
        { new: true }
      ).exec();
      return result;
    } catch (error) {
      throw new Error(`Failed to update chat pin status: ${error.message}`);
    }
  }
} 